File: src/main/scala/kim/dencka/scala/easy/KidsWithTheGreatestNumberOfCandies.scala
package kim.dencka.scala.easy

object KidsWithTheGreatestNumberOfCandies extends App {

  def kidsWithCandies(candies: Array[Int], extraCandies: Int): List[Boolean] = {
    val max    = candies.max
    val result = Array.ofDim[Boolean](candies.length)
    result.indices.foreach { i =>
      if (max <= candies(i) + extraCandies) result(i) = true
    }
    result.toList
  }

  println(kidsWithCandies(Array(2, 3, 5, 1, 3), 3)) // output: List(true,true,true,false,true)
  println(kidsWithCandies(Array(4, 2, 1, 1, 2), 1)) // output: List(true,false,false,false,false)

}

File: src/main/scala/kim/dencka/java/easy/KidsWithTheGreatestNumberOfCandies.java
package kim.dencka.java.easy;

import java.util.ArrayList;
import java.util.List;

public class KidsWithTheGreatestNumberOfCandies {
    public static void main(String[] args) {
        int[] candies = {2, 3, 5, 1, 3};
        int extraCandies = 3;

        System.out.println(kidsWithCandies(candies, extraCandies));

    }

    public static List<Boolean> kidsWithCandies(int[] candies, int extraCandies) {
        List<Boolean> result = new ArrayList<>();
        int max = Integer.MIN_VALUE;
        for (int num : candies) {
            max = Math.max(max, num);
        }
        for (int num : candies) {
            result.add(num + extraCandies >= max);
        }
        return result;
    }
}
